src/worker/ingestion/hooks.ts

```diff
@@ -0,0 +1,225 @@
+    const matchedTokens = messageFormat.match(/(?<=\[)(.*?)(?=\])/g) || []
+    let tokenizedMessage = messageFormat
+    if (matchedTokens.length) {
+        tokenizedMessage = tokenizedMessage.replace(/\[(.*?)\]/g, '%s')
+    }
+    return [matchedTokens, tokenizedMessage]
+}
+
+export function getValueOfToken(
+    action: Action,
+    event: PluginEvent,
+    person: Person | undefined,
+    siteUrl: string,
+    webhookType: WebhookType,
+    tokenParts: string[]
+): [string, string] {
+    let text = ''
+    let markdown = ''
+
+    if (tokenParts[0] === 'user') {
+        if (tokenParts[1] == 'name') {
+            ;[text, markdown] = getUserDetails(event, person, siteUrl, webhookType)
+        } else {
+            const propertyName = `$${tokenParts[1]}`
+            const property = event.properties?.[propertyName]
+            text = markdown = stringify(property)
+        }
+    } else if (tokenParts[0] === 'action') {
+        if (tokenParts[1] == 'name') {
+            ;[text, markdown] = getActionDetails(action, siteUrl, webhookType)
+        }
+    } else if (tokenParts[0] == 'event') {
+        if (tokenParts[1] == 'name') {
+            text = markdown = stringify(event.event)
+        } else if (tokenParts[1] == 'properties' && tokenParts.length > 2) {
+            const propertyName = tokenParts[2]
+            const property = event.properties?.[propertyName]
+            text = markdown = stringify(property)
+        }
+    } else {
+        throw new Error()
+    }
+    return [text, markdown]
+}
+
+export function getFormattedMessage(
+    action: Action,
+    event: PluginEvent,
+    person: Person | undefined,
+    siteUrl: string,
+    webhookType: WebhookType
+): [string, string] {
+    const messageFormat = action.slack_message_format || '[action.name] was triggered by [user.name]'
+    let messageText: string
+    let messageMarkdown: string
+
+    try {
+        const [tokens, tokenizedMessage] = getTokens(messageFormat)
+        const values: string[] = []
+        const markdownValues: string[] = []
+
+        for (const token of tokens) {
+            const tokenParts = token.match(/\w+/g) || []
+
+            const [value, markdownValue] = getValueOfToken(action, event, person, siteUrl, webhookType, tokenParts)
+            values.push(value)
+            markdownValues.push(markdownValue)
+        }
+        messageText = format(tokenizedMessage, ...values)
+        messageMarkdown = format(tokenizedMessage, ...markdownValues)
+    } catch (error) {
+        const [actionName, actionMarkdown] = getActionDetails(action, siteUrl, webhookType)
+        messageText = `⚠ Error: There are one or more formatting errors in the message template for action "${actionName}".`
+        messageMarkdown = `*⚠ Error: There are one or more formatting errors in the message template for action "${actionMarkdown}".*`
+    }
+
+    return [messageText, messageMarkdown]
+}
+
+export class HookCannon {
```

Comment:
what is `Cannon` supposed to imply here? (a bit lost with the naming 😅 )

Is it because it finds **and fires** hooks?

Reply:
Literally yes, it definitely may be too "fun" LOL. Feel free to propose something clearer

Reply:
As someone who called a thread-killer the `Terminator`, yes, I'm very happy with a hook cannon 😂 .

Although `hookCannon.findAndFireHooks` feels like it's a `HookCannonAndLogistician`.

To make things clearer, perhaps a `HookManager`? It seems to fit in well with our other managers, and performs a similar role to the others (although they fetch, while this one fires)

Reply:
Renamed to, drum roll, `HookCommander`

Reply:
Ten-Hut!